Make error more explicit when PKI dir not present for salt-call

terminalmage opened the pull request:
Refs: #40948

terminalmage committed:
Make error more explicit when PKI dir not present for salt-call

salt-jenkins commented:
@terminalmage, thanks for your PR! By analyzing the history of the files in this pull request, we identified @skizunov, @jacksontj and @thatch45 to be potential reviewers.

terminalmage commented:
Not sure at this point if this is the correct fix.

cachedout commented:
@terminalmage Please let me know when you have an update on this. Thanks.

terminalmage commented:
I commented on the original issue. I was waiting for confirmation from the person that filed the initial issue, confirmation that never came. I personally was not able to reproduce this, but I only tested on Linux and did not have a MacOS box available for testing.

cachedout commented:
I'm just going to close this then. If new info comes back and we need to re-open it, so be it.

ScoreUnder commented:
Sorry, I managed to overlook this one.

I don't have a freshly set-up OSX machine to hand yet (I could commandeer one of our staging ones, but I don't know if there is much point yet), so I have instead just erased the `/etc/salt/pki` directory on a free development machine (Linux) for testing purposes.

With the current patch:
```
~/src/salt (git)-[remotes/terminalmage/issue40948] % sudo ./saltenv/bin/salt-call test.ping -l debug
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/masters.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/masters.conf
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: {my hostname}
[DEBUG   ] Configuration file path: /etc/salt/minion
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/masters.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/masters.conf
[DEBUG   ] Please install 'virt-what' to improve results of the 'virtual' grain.
[DEBUG   ] Connecting to master. Attempt 1 of 1
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', '{my hostname}', 'tcp://{master 1}:4506')
[INFO    ] Master {master 1} could not be reached, trying next master (if any)
[WARNING ] Master ip address changed from {master 1} to {master 2}
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', '{my hostname}', 'tcp://{master 2}:4506')
[INFO    ] Master {master 2} could not be reached, trying next master (if any)
[ERROR   ] No master could be reached or all masters denied the minions connection attempt.
Could not access /etc/salt/pki. Path does not exist.
```
It still shows the spurious "could not be reached" error.

With a patch for a little more info:
```diff
diff --git a/salt/minion.py b/salt/minion.py
index 8b1537d9b7..9d6fa2b176 100644
--- a/salt/minion.py
+++ b/salt/minion.py
@@ -555,13 +555,15 @@ class MinionBase(object):
                         break
                     except SaltClientError as exc:
                         last_exc = exc
-                        if exc.strerror.startswith(
-                                'Could not access {0}'.format(opts['pki_dir'])):
+                        startswith_str = 'Could not access {0}'.format(opts['pki_dir'])
+                        if exc.strerror.startswith(startswith_str):
                             msg = exc.strerror
                         else:
                             msg = ('Master {0} could not be reached, trying '
                                    'next master (if any)'.format(opts['master']))
                         log.info(msg)
+                        log.info("type(exc)=%r, exc.strerror=%r" % (type(exc), exc.strerror))
+                        log.info("exc.strerror.startswith(%r)=%r" % (startswith_str, exc.strerror.startswith(startswith_str)))
                         continue
 
                 if not conn:
```

```
~/src/salt (git)-[remotes/terminalmage/issue40948] % sudo ./saltenv/bin/salt-call test.ping -l debug
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/masters.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/masters.conf
[DEBUG   ] Using cached minion ID from /etc/salt/minion_id: {my hostname}
[DEBUG   ] Configuration file path: /etc/salt/minion
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] Reading configuration from /etc/salt/minion
[DEBUG   ] Including configuration from '/etc/salt/minion.d/masters.conf'
[DEBUG   ] Reading configuration from /etc/salt/minion.d/masters.conf
[DEBUG   ] Please install 'virt-what' to improve results of the 'virtual' grain.
[DEBUG   ] Connecting to master. Attempt 1 of 1
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', '{my hostname}', 'tcp://{master 1}:4506')
[INFO    ] Master {master 1} could not be reached, trying next master (if any)
[INFO    ] type(exc)=<class 'salt.exceptions.SaltClientError'>, exc.strerror='Could not access /etc/salt/pki. Path does not exist.'
[INFO    ] exc.strerror.startswith('Could not access /etc/salt/pki/minion')=False
[WARNING ] Master ip address changed from {master 1} to {master 2}
[DEBUG   ] Initializing new AsyncAuth for ('/etc/salt/pki/minion', '{my hostname}', 'tcp://{master 2}:4506')
[INFO    ] Master {master 2} could not be reached, trying next master (if any)
[INFO    ] type(exc)=<class 'salt.exceptions.SaltClientError'>, exc.strerror='Could not access /etc/salt/pki. Path does not exist.'
[INFO    ] exc.strerror.startswith('Could not access /etc/salt/pki/minion')=False
[ERROR   ] No master could be reached or all masters denied the minions connection attempt.
Could not access /etc/salt/pki. Path does not exist.
```

It looks like the path is too specific. Also, the error saying "No master could be reached or all masters denied the minions connection attempt" will still exist at the end.

cachedout commented:
Base branch changed to 2016.11

terminalmage committed:
Fix path in log message

The actual error refers to the parent dir of the master/minion/etc. pki
subdir (i.e. /etc/salt/pki instead of /etc/salt/pki/minion). Therefore,
the startswith check needs to look for the parent of the pki_dir.

terminalmage committed:
Make error more specific to aid in troubleshooting

terminalmage commented:
@ScoreUnder Thanks for the heads-up on the `pki_dir`, I have updated the log message to properly identify these log messages.

I personally think that the log message that happens at the end is still accurate, the minion can't reach any of the masters if the permissions forbid it.

rallytime commented on salt/minion.py:
```diff
@@ -567,8 +567,16 @@ def eval_master(self,
-                        msg = ('Master {0} could not be reached, trying '
-                               'next master (if any)'.format(opts['master']))
+                        pki_dirname = os.path.dirname(opts['pki_dir'])
+                        if exc.strerror.startswith(
```
Is `startswith` the best way to check for this? Or, rather, would a smaller portion of the message make more sense to check for?

terminalmage replied:
@rallytime yes, we do need to use startswith because we need to check that the message starts with `Could not access <path>`. But, the logic there was incorrect, I was doing the startswith on the message I wanted to log. D'oh!

I've updated and made the startswith more general so we aren't just looking at the PKI dir.

terminalmage committed:
Fix incorrect logic in exception check
